### ISSUE TITLE
Conversation With the Model

### ISSUE THREAD

**Issue by Mark-Phillipson** [C-Loftus/talon-ai-tools#78]
Would be nice to be able to have a conversation with the model with the responses back in voice.

For example "what is the capital of France?" and the model would reply with "Paris".

Talon file script example:

```talon
computer <user.text>:
    # will take the argument as a question and reply with voice
    text = user.text
    result = user.gpt_apply_prompt("Please reply in summary only", text)
    user.tts(result)
```

This is a simple example of how to use the model to reply to a question in voice. The model will reply with a summary of the text.

It would also be beneficial to have a feature that allows us to add each interaction to a list for the model to record. Additionally, we would require the functionality to clear this list when necessary.

For example "what is the capital of France?" and the model were reply it with "Paris". Then the user asked "what about Germany?" The model would reply with "Berlin".

```talon
computer <user.text>:
    # will take the argument as a question and reply with voice
    text = user.text
    user.add_to_list(text)
    result = user.gpt_apply_prompt("Please reply in summary only", user.get_list())
    user.tts(result)
```

This is a simple example of how to use the model to reply to a list of questions in voice. The model will reply with a summary of the list.  Note the add to list functionality would have to be implemented.

Also it's a bit of a stretch to have the text to speech here as not everybody would have it installed so not sure how to get around that.

**Comment by jaresty** [comment]
I think tts could be another destination-wdyt, @C-Loftus ?  You could say "model <prompt > spoken" and have it reply with tts.

**Comment by C-Loftus** [comment]
Thanks for your comment 

1. With regards to the point on conversion, yes that is something I would like to support. Currently all requests are stateless. If we want to allow state we need:
      *  A way to make sure the state of the conversation has a very clear start and end. (i.e. we don't want the user to accidentally include irrelevant state to the next request) 
      * A gui that is at least optional, to go along with the request. (i.e. like the one you'd see with chatgpt, but more minimal. Just to clearly represent the state of the screen). 
      * A way to make it feel consistent and natural even while using the stateless commands. 
          * I don't want to start fragmenting the grammar or start to add fragmenting keywords, so I need to brainstorm on this a bit. Probably good to combine this with https://github.com/C-Loftus/talon-ai-tools/issues/74 in some way

2. @Mark-Phillipson with regards to TTS, have you used any of the commands in [the TTS folder](https://github.com/C-Loftus/talon-ai-tools/blob/main/TTS/gpt-tts.talon)? (i.e. do you have https://github.com/C-Loftus/sight-free-talon installed for TTS?) I am curious on your user experience and if you find it useful. I also think that tts is very useful and makes it so you don't need to clutter the screen or paste anything. However, installing tts is a pain at the moment since Talon doesn't have a package manager.
     * @jaresty I agree that something like "echoed" or "spoken" could be a good `insertionDestination`. So the grammar would end up being something like `model fix grammar echoed` or `model fix grammar spoken`. So we keep the point free / pipeline pattern. 
     * I could exploring splitting sight-free-talon to be one repo that is just the tts library and another that is the code for screen reader users. And then we could bundle the tts lib as a git submodule. But this might be overkill. Not sure how annoying people find sight-free-talon installation to be.

**Comment by Mark-Phillipson** [comment]
Yes I had to install the tts manually which was not straight forward.

Sometimes whilst working I'm trying to remember the name of for example a CSS property and justice need a quick reminder and don't want to lose my place or focus. So I thought it would be cool to have a text to speech feature that I can trigger with a voice command.

Another example when I'm reading an article on a website I can say computer define whatever the word is that I need a definition for without losing my place.

**Comment by C-Loftus** [comment]
> Yes I had to install the tts manually which was not straight forward.

When you say manually do you mean you copied and pasted just detect to speech code from sight-free-talon and didn't clone that repository, or rather than you cloned sight-free-talon but didn't find it intuitive to use? 

Do you have any suggestions on how I could make it more intuitive? 

> Sometimes whilst working I'm trying to remember the name of for example a CSS property and just need a quick reminder and don't want to lose my place or focus. So I thought it would be cool to have a text to speech feature that I can trigger with a voice command.

Yup I agree!

**Comment by Mark-Phillipson** [comment]
I did clone the site free talon repository but as I only did it once I can't really remember much about it suffice to say it is working and I can remember having to change the speed of the voice to be able to understand it.

**Comment by C-Loftus** [comment]
Should be implemented now. You can do the following to have a conversation with the model which can be optionally verbal via TTS if you would like

1. `model start thread` -> which will store your conversation in a new thread
2. `model toggle window` -> to open the window which shows thread visualizations
3. Any model command you request (i.e. `model please tell me X`) will be auto added to the window after it returns its result
4. If you use `to speech` as your model destination, it will speak the output and if you have the thread enabled, it will continue to update the window.
    * You can override the default destination to be speech if  you always want speech. We have a setting for that now.

If you feel this is missing behavior, please file a new issue so we can iterate

**Comment by Mark-Phillipson** [comment]
After updating today now getting the following and no response when I say "Model Apply Clip":

```
talon.scripting.talon_script.TalonScriptError: 
 in script at c:\users\mphil\appdata\roaming\talon\user\talon-ai-tools\GPT\gpt.talon:20:
   prompt = clip.text()
   text = edit.selected_text()
   result = user.gpt_apply_prompt(prompt, text)
 > user.paste(result)
TypeError: 'dict' object cannot be converted to 'PyString'
```

**Comment by atljoseph** [comment]
GPT Failure: text to speech is not installed.

case "textToSpeech":
                try:
                    actions.user.tts(message_text_no_images)
                except KeyError:
                    notify("GPT Failure: text to speech is not installed")

To be fair, the TTS usage here is _very_ cryptic.

So, `actions.user.tts` is a ... function?? That function doesn't exist in this repo.

No doc says to go look at the other repo. Searched this entire repo for regex: `tts|textToSpeech`. About 4-5 instances. All code. No helpful comments.

Based on the responses above, it seems like something "is implemented" differently than before.
- Still, don't know what i don't know, and the repo doesn't say anything in regards to TTS.

---

Also, the threading thing is a bit confusing. I like the idea, but I'm missing a lot here.
- Where do these threads live? 
- How can i discover them or manage them (other than start, stop, clear)?
- The confirmation GUI NEVER shows ANY text... only buttons. Something is broken there.

---

More explanation and examples are needed to understand how context and threading works. Right now, its pretty intimidating.

**Comment by jaresty** [comment]
I can't speak to the tts support but the threading is entirely in memory. The entire conversation is sent with every request for threads and visualization of that is rather difficult using existing tools in talon so honestly I don't use it that much. I'm a much more active user of context because I can control it very carefully. Also there are plenty of tools available that handle threading pretty well for you already so you might try one of those 🤷

**Comment by atljoseph** [comment]
Yeah, thanks for the comments. I’m just asking about this library only. I’m
aware of other things that are out there. Thank you.

**Comment by atljoseph** [comment]
Is the context stored in a file? Which file? Is there a command to open the context to see it? I thought that was what the confirmation GUI is for, but it doesn’t show any text from the model. Only three buttons something is broken.

**Comment by jaresty** [comment]
The context is also in memory. You can visualize it by using the command 'pass'. You can try using 'model pass context' to have the context inserted into the document that you are currently in.

**Comment by atljoseph** [comment]
Ok. Yeah that doesn’t work for me. File system is memory too lol

**Comment by jaresty** [comment]
It's not on the file system it's just in array inside the process. You can also use destinations to display the context. for example you can use 'model pass context to window' to display the context in a window.

**Comment by jaresty** [comment]
With respect to having multiple contexts yeah I think that could be useful but I'm not maintaining this repository myself. As it's easy to put things into and out of context I think you may consider just keeping a journal and putting those things into folders and bringing them back when you need them. at least that's what I was thinking of doing when I thought I would use that more. In my own practice I've found that I am often throwing away context and creating them for one off problems so I don't really need to save them. I'm offering some suggestions for help here but not really offering much in the way of support since I don't maintain this repository right now.

**Comment by atljoseph** [comment]
Ok thank you for the suggestions. I’ll wait on Colton to reply, since it’s
his repository, right now.

**Comment by C-Loftus** [comment]
TTS is provided by https://github.com/C-Loftus/sight-free-talon 

In general I have focused on keeping this repo stable for the time being since it was getting to much of a maintenance burden for me. I am hoping that Talon will add a package manager / testing system / improved GUI tools that will make some of this work easier.  If you feel there are docs or behavior missing, feel free to submit a PR and I would be happy to look at it. I agree that more work needs to be done documenting things. I would love to do that but unfortunately just don't have the bandwidth.
